File: 104.二叉树的最大深度.cpp
/*
 * @lc app=leetcode.cn id=104 lang=cpp
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	int searchTree(TreeNode* t, int d)
	{
		if (t->left == nullptr && t->right == nullptr) return d;
		int d1 = d, d2 = d;
		if (t->left != nullptr)
		{
			d1 = searchTree(t->left, d + 1);
		}
		if (t->right != nullptr)
		{
			d2 = searchTree(t->right, d + 1);
		}
		return d1 > d2 ? d1 : d2;
	}
    int maxDepth(TreeNode* root) {
		if (root == nullptr) return 0;
		return searchTree(root, 1);
    }
};
// @lc code=end

File: Answer/104.二叉树的最大深度.cs
/*
 * @lc app=leetcode.cn id=104 lang=csharp
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int MaxDepth(TreeNode root) {
        if(root == null) return 0;
        Queue<TreeNode> tree = new Queue<TreeNode>{};
        Queue<int> deep = new Queue<int>{};
        tree.Enqueue(root);
        deep.Enqueue(1);
        TreeNode t;
        int d = 1;
        while(tree.Count != 0)
        {
            t = tree.Dequeue();
            d = deep.Dequeue();
            if(t.left != null)
            {
                tree.Enqueue(t.left);
                deep.Enqueue(d + 1);
            }
            if(t.right != null)
            {
                tree.Enqueue(t.right);
                deep.Enqueue(d + 1);
            }
        }
        return d;
    }
}
// @lc code=end

